classic_tetris_project/facades/tournament_match_display.py

```diff
@@ -52,3 +54,63 @@ def player2_display_name(self):
         else:
             return self.display_name_from_source(self.tournament_match.source2_type,
                                                  self.tournament_match.source2_data)
+
+    def player1_color(self):
+        return self.get_color_for_player_index(self.tournament_match, 1)
+
+
+    def player2_color(self):
+        return self.get_color_for_player_index(self.tournament_match, 0)
+
+    @staticmethod
+    def generate_round(match_count, all_matches):
+        new_matches = []
+        matches_to_generate = match_count if len(all_matches) >= match_count else len(all_matches)
+        for _ in range(matches_to_generate):
+            next_match = {}
+            if len(all_matches) != 0:
+                next_match = all_matches.pop(0)
+            new_matches.append(next_match)
+        
+        new_matches.reverse()
+        return new_matches
+
+    @staticmethod
+    def get_color_for_player_index(match, player_position):
```

Comment:
The pattern i see is that you only need player_index and total_player_count (rounded up to nearest power of 2) to figure out which of the four brackets they are in.  So i think this can be greatly simplified.

Reply:
Here you go:
```py
import math
from collections import OrderedDict
def branch(seed, level, limit, results):
    """
    Returns list of players in order of play.
    you can change results.append into a pair if you want
    """
    levelSum = 2 ** level + 1
    if (limit == level + 1):
        results.append(seed)
        results.append(levelSum - seed)        
        return
    elif (seed % 2 == 1):
        branch(seed, level + 1, limit, results);
        branch(levelSum - seed, level + 1, limit, results)
    else:
        branch(levelSum - seed, level + 1, limit, results)
        branch(seed, level + 1, limit, results)

def get_seed_order(num_teams, num_colors):
    """
    Given a po2 num_teams, and num_colors that evenly divides,
    returns a dictionary that maps seed number to bracket color.
    """
    #num_teams must be a power of 2
    limit = math.log(num_teams, 2) + 1
    results = []
    branch(1, 1, limit, results)
    
    step = len(results) // num_colors
    seeds = OrderedDict()
    for index, item in enumerate(results):
        seeds[item] = (index // step) + 1
        
    return seeds

NUM_COLORS = 4
#Use below if you want to access via num_players followed by seed number
BRACKET_COLORS = { 2**x: get_seed_order(2**x,NUM_COLORS) for x in range(3,8)}
#Use below if you want seed number only. It's still correct-ish.
BRACKET_COLORS = get_seed_order(2**8,NUM_COLORS)
if __name__ == '__main__':
    print(BRACKET_COLORS)    
```
This generates pre-calculated colors based on the players seed. You still have to use the "grey" code when you have non power of two colors, but you can now lookup player by going `BRACKET_COLORS[NUM_PLAYERS_PO2][seed]`

I think that this can be further simplified by a reverse reverse sweep (iterate 1 to 4, then 8 to 5, then 9 to 16, then 32 to 17, incrementing color by 1 each time

Reply:
Hi @alex-ong thanks for the suggestion!  This is definitely a way simpler approach and the logic checks out for me, thanks for writing it out!

I'll have to circle back tomorrow after some sleep, but one potential thought on this idea is that I don't think it would skin matches that don't currently have seeded players, e.g. "Winner of Match X" blocks do not know what seed they should be.

I vastly prefer your simpler approach though, so perhaps there is a way to resolve that caveat.  Through my complicated mess this specific issue is solved because it calculates an "index" based on the round and match number.

And also I'm wondering if there is extra logic that would be needed for the "feed in" matches.  Screenshot example of your implementation here for a non power of 2 bracket:

![Screenshot 2022-03-05 015805](https://user-images.githubusercontent.com/1138903/156876326-f3b3712a-4f25-415d-b76a-0f0ac3a7be1d.png)

I'll revisit tomorrow though and take another look since I really think this concept is cleaner.

Reply:
Looking at this with some fresh eyes has made me realize that basically all the problems go away with a proper tree structure, so I'm gonna give that a try.  Should clear a lot of things up.

Reply:
Herp i didnt think about the unplayed future matches. If you create a tree like you said you can potentially mark all decendent children as being in the same color i guess? Hmm more thinking

Reply:
@alex-ong I came up with a solution here that I will share back since the changes are in the submodule.  There could be some more clean up opportunity, but I am MUCH happier with this tree based approach.  Thank you for your thoughts regardless!  Curious what you think too:

```python
class MatchNode:
    def __init__(self, match):
        self.left = None
        self.right = None
        self.match = match
        self.color_left = 5
        self.color_right = 5

    def color_all(self, color_index):
        self.color_left = color_index
        self.color_right = color_index

        if self.left:
            self.left.color_all(color_index)

        if self.right:
            self.right.color_all(color_index)

    def get_nodes_at_round(self, round_num):
        nodes = [self]
        for i in range(round_num - 1):
            children = []
            for node in nodes:
                children.extend([node.left, node.right])
            nodes = children
        return nodes
```

```python
        bracket_nodes = []
        for match in all_matches:
            node = MatchNode(match)
            bracket_nodes.append(node)

        for node in reversed(bracket_nodes):
            left_source = node.match.tournament_match.source1_type
            left_data = node.match.tournament_match.source1_data
            right_source = node.match.tournament_match.source2_type
            right_data = node.match.tournament_match.source2_data

            if left_source == TournamentMatch.PlayerSource.MATCH_WINNER:
                node.left = bracket_nodes[left_data - 1]

            if right_source == TournamentMatch.PlayerSource.MATCH_WINNER:
                node.right = bracket_nodes[right_data - 1]

        root_bracket_node = bracket_nodes[-1]
        semi_left = root_bracket_node.left
        semi_right = root_bracket_node.right
        semi_left.color_left = 1
        semi_left.color_right = 2
        semi_right.color_left = 3
        semi_right.color_right = 4

        semi_left.left.color_all(1)
        semi_left.right.color_all(2)
        semi_right.left.color_all(3)
        semi_right.right.color_all(4)

        bracket_matches = []

        for i in range(root_bracket_node.match.tournament_match.round_number, 0, -1):
            matches_at_round = root_bracket_node.get_nodes_at_round(i)
            bracket_matches.append(matches_at_round)
```

And with that... boom:

![Screenshot 2022-03-05 160119](https://user-images.githubusercontent.com/1138903/156902677-524a96df-e442-4e9a-8c72-e9d14d33b975.png)

Reply:
Very nice! Might want to be a bit more modular, instead of semi_left / semi_right, have a function that given a Po2 number of colors, works backwards until it hits enough matches (2 for semi, 4 for quarters), that way in future its easy to expand it to 8 colors if need be.

```py
root_bracket_node = bracket_nodes[-1]
semi_left = root_bracket_node.left
semi_right = root_bracket_node.right
semi_left.color_left = 1
semi_left.color_right = 2
semi_right.color_left = 3
semi_right.color_right = 4

semi_left.left.color_all(1)
semi_left.right.color_all(2)
semi_right.left.color_all(3)
semi_right.right.color_all(4)
```

could be replaced by
```py
#pseudocode
TARGET_LEVEL = NUM_COLORS/2
nodes = []
get_level(root, TARGET_LEVEL, nodes) #returns nodes in that level
for i, node in enumerate(nodes):
    node.left.color_all(2*i+1)
    node.right.color_all(2*i+2)
    
#untested
def get_level(root, level, nodes, current_level = 1):
    if node is None:
        return
    if level == current_level:
        nodes.append(root)
    else:
        get_level(root.left, level, nodes, current_level+1)
        get_level(root.right, level, nodes, current_level+1)
```